runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/ProcessRemoteBundleOperation.java

```diff
@@ -100,6 +197,75 @@ public void finish() throws Exception {
       } catch (Exception e) {
         throw new RuntimeException("Failed to finish remote bundle", e);
       }
+
+      // TODO(BEAM-6274): do we have to put this in the "start" method as well?
+      try (RemoteBundle bundle =
+          stageBundleFactory.getBundle(receiverFactory, stateRequestHandler, progressHandler)) {
+
+        // TODO(BEAM-6274): Why do we need to namespace this to "user"?
+        DataflowExecutionContext.DataflowStepContext stepContext =
+            executionContext
+                .getStepContext((DataflowOperationContext) this.context)
+                .namespacedToUser();
+
+        // TODO(BEAM-6274): investigate if this is the correct window
+        TimerInternals.TimerData timerData =
+            stepContext.getNextFiredTimer(GlobalWindow.Coder.INSTANCE);
+        while (timerData != null) {
+          LOG.debug("Found fired timer in start {}", timerData);
+
+          // TODO(BEAM-6274): get the correct payload and payload coder
+          StateNamespaces.WindowNamespace windowNamespace =
+              (StateNamespaces.WindowNamespace) timerData.getNamespace();
+          BoundedWindow window = windowNamespace.getWindow();
+
+          WindowedValue<KV<Object, Timer>> timerValue =
+              WindowedValue.of(
+                  KV.of(
+                      timerIdToKey.get(timerData.getTimerId()),
+                      Timer.of(timerData.getTimestamp(), new byte[0])),
+                  timerData.getTimestamp(),
+                  Collections.singleton(window),
+                  PaneInfo.NO_FIRING);
+
+          String mainInputId =
+              timerIdToTimerSpecMap.get(timerData.getTimerId()).inputCollectionId();
+
+          bundle.getInputReceivers().get(mainInputId).accept(timerValue);
+
+          // TODO(BEAM-6274): investigate if this is the correct window
+          timerData = stepContext.getNextFiredTimer(GlobalWindow.Coder.INSTANCE);
+        }
+      }
+    }
+  }
+
+  private void receive(String pCollectionId, Object receivedElement) throws Exception {
+    LOG.debug("Received element {} for pcollection {}", receivedElement, pCollectionId);
+    // TODO(BEAM-6274): move this out into its own receiver class
```

Comment:
I'm not sure if relevant ticket covers it based on description, skip if it is.
I feel that it is possible to move most of code that is added to a separate class with clear responsibility of tracking time.
It would be really great if we can do it. 
It should simplify unittesting and add cleaner boundary between functionality.

Reply:
I plan on a follow up PR cleaning this up and moving into its own class.